include/envoy/stats/refcount_ptr.h

```diff
@@ -0,0 +1,153 @@
+#pragma once
+
+#include <atomic>
+
+#include "envoy/common/pure.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Stats {
+
+// Implements a reference-counted pointer to a class, so that its external usage
+// model is identical to std::shared_ptr, but the reference count itself is held
+// in the class. The class is expected to implement three methods:
+//    void incRefCount()
+//    bool decRefCount()  -- returns true if the reference count goes to zero.
+//    uint32_t use_count()
+// It may implement them by delegating to RefcountHelper (see below), or by
+// inheriting from RefcountInterface (see below).
+template <class T> class RefcountPtr {
+public:
+  RefcountPtr() : ptr_(nullptr) {}
+
+  // Constructing a reference-counted object from a pointer; this is safe to
+  // do when the reference-count is held in the object. For example, this code
+  // crashes:
+  //   {
+  //     std::shared_ptr<std::string> a = std::make_shared<std::string>("x");
+  //     std::shared_ptr<std::string> b(a.get());
+  //   }
+  // whereas the analogous code for RefcountPtr works fine.
+  RefcountPtr(T* ptr) : ptr_(ptr) {
+    if (ptr_ != nullptr) {
+      ptr_->incRefCount();
+    }
+  }
+
+  RefcountPtr(const RefcountPtr& src) { set(src.get()); }
```

Comment:
For this ctor and below, I think it's preferable to delegate to the above ctor, like:
```cc
RefcountPtr(const RefcountPtr& src) : RefcountPtr(src.get) {}
```